### PR TITLE
[dv] Move sim_tops to {tool}.hjson

### LINKED ISSUE #4340: dvsim: QSTA support - sim_tops needs to be moved to run_opts
Hello,
  I now have a compile-ready, sim-starting Questa in dvsim flow :-)

I had to change the **build_opts** inside: **common_sim_cfg.hjson** 

This is so because, in Questa 2-command flow, vlog-vsim, vlog is a compiler and not an elaborator. So top level needs to be passed to vsim and not vlog. Going by dvsim's build_opts, I had to move the {sim_opts} to run_opts. Code snippet below.

Now - what do you folks suggest - add a separate build_opts_qsta - sounds like an ugly work-around to me, but to keep minimal changes to existing flow maybe. Ideally we perhaps need to add a common_build_opts and VCS/XLM/QSTA/RVRA build_opts - thinking loud, maybe there are other/better solutions.

Thanks
Srini

```json

// Default build and run opts
  build_opts: [// List multiple tops for the simulation
               // VW TBD - maybe we need build_opts_qsta "{sim_tops}",
               // Standard UVM defines
```

My changes are now pushed to a local fork, will share soon.

### PR DESCRIPTION
**Pull request by sriyerg**
Fixes #4340. 

This change factors `sim_tops` out of `common_sim_cfg.hjson` in favor of being placed in `{tool}.hjson` instead. This is done to support Questa (support for which will be added soon). 

The first commit fixes a bug in DVSim (prevent `${..}` constuct from being treated as a substitution var). 

The second moves `sim_tops` expansion out of the common hjson and moves it into individual tool-specific hjson. How tops are to be specified is also changed - there is no need to add `-top` prefix to each top level. This is factored in the tool hjson already.

### CODE REVIEW

File: util/dvsim/utils.py [hunk 1]
```diff
@@ -116,7 +116,7 @@ def _subst_wildcards(var, mdict, ignored, ignore_error, seen):
     and seen_err is true if we stopped early because of an ignored error.
 
     '''
-    wildcard_re = re.compile(r"{([A-Za-z0-9\_]+)}")
+    wildcard_re = re.compile(r"(?<!\$){([A-Za-z0-9\_]+)}")
```
Comment (rswarbrick) [inline util/dvsim/utils.py:119]: Hmm, ok, but this is a bit magical. I wonder whether it would be better to have some sort of explicit escaping (like Python's format function). So shell code that absolutely needed the braces (unusual) would end up looking like `${{foo}}`.

File: hw/dv/tools/dvsim/dsim.hjson [hunk 2]
```diff
@@ -23,6 +23,8 @@
                "-c-opts -I{DSIM_HOME}/include",
                "-timescale 1ns/1ps",
                "-f {sv_flist}",
+               // List multiple tops for the simulation. Prepend each top level with `-top`.
+               '''{eval_cmd} tops=(`echo {sim_tops}`); echo "${tops[@]/#/-top }"''',
```
Comment (rswarbrick) [inline hw/dv/tools/dvsim/dsim.hjson:27]: While we discuss the escaping in the previous commit, you could do this with sed (and no escaping required):
```
echo {sim_tops} | sed -E 's/(\w+)/-top \1/g'
```
Maybe split out the previous commit into a separate PR, and we can figure out the best approach for that separately from this change?

File: hw/dv/tools/dvsim/vcs.hjson [hunk 3]
```diff
@@ -15,6 +15,8 @@
                "-Mdir={build_ex}.csrc",
                "-o {build_ex}",
                "-f {sv_flist}",
+               // List multiple tops for the simulation. Prepend each top level with `-top`.
+               '''{eval_cmd} tops=(`echo {sim_tops}`); echo "${tops[@]/#/-top }"''',
```
Comment (rswarbrick) [inline hw/dv/tools/dvsim/vcs.hjson:19]: We've got two `{eval_cmd}` lines in this file that both compute the same thing. Maybe put it in a separate variable to avoid the unneeded extra subprocess?

**Review by rswarbrick — CHANGES_REQUESTED** [review]
I think we need to discuss the escaping change a bit more carefully, but my `sed` suggestion should decouple that from getting things working for Questa. Maybe split up the PR to get the Questa side fixed quickly?

**Review by weicaiyang — APPROVED** [review]
LGTM

**Review by sriyerg — COMMENTED** [review]
Thanks, I dropped the first commit and replaced the echo with sed.

**Review by weicaiyang — APPROVED** [review]

**Review by udinator — APPROVED** [review]

File: hw/dv/tools/dvsim/vcs.hjson (hunk 3 above)
Comment (rswarbrick) [reply hw/dv/tools/dvsim/vcs.hjson]: Duh, I didn't read this carefully enough! You're completely right.

**Review by rswarbrick — APPROVED** [review]
Looks good to me, thanks!
